Sources/ClientRuntime/Networking/Http/URLSession/FoundationStreamBridge.swift

```diff
@@ -44,80 +50,99 @@ class FoundationStreamBridge: NSObject, StreamDelegate {
+        func perform(_ block: @escaping @Sendable () async throws -> Void) async throws {
+            let task = Task { [task] in
                 _ = await task?.result
-                try await block(self)
+                try await block()
             }
+            self.task = task
+            _ = try await task.value
         }
     }
 
     /// Actor used to enforce the order of multiple concurrent stream writes.
     private let writeCoordinator = WriteCoordinator()
 
-    /// A shared serial DispatchQueue to run the stream operations.
-    /// Performing operations on an async queue allows Swift concurrency tasks to not block.
+    /// A serial `DispatchQueue` to run the stream operations for the Foundation `OutputStream`.
+    ///
+    /// Operations performed on the queue include:
+    /// - Opening the stream
+    /// - Closing the stream
+    /// - Writing to the stream
+    /// - Receiving `StreamDelegate` callbacks
+    ///
+    /// Queue operations are run in the order they are started, but not necessarily exclusively.
     private let queue = DispatchQueue(label: "AWSFoundationStreamBridge")
 
+    /// `true` if the readable stream has been closed, `false` otherwise.  Will be flipped to `true` once the readable stream is read,
+    /// and `nil` is returned.
+    ///
+    /// Access this variable only during a write operation to ensure exclusive access.
+    private var readableStreamIsClosed = false
+
     // MARK: - init & deinit
 
-    /// Creates a stream bridge taking the passed `ReadableStream` as its input
+    /// Creates a stream bridge taking the passed `ReadableStream` as its input., and exposing a Foundation `InputStream`
+    /// that may be used for streaming data on to Foundation components.
     ///
     /// Data will be buffered in an internal, in-memory buffer.  The Foundation `InputStream` that exposes `readableStream`
     /// is exposed by the `inputStream` property after creation.
     /// - Parameters:
     ///   - readableStream: The `ReadableStream` that serves as the input to the bridge.
-    ///   - bufferSize: The number of bytes in the in-memory buffer.  The buffer is allocated for this size no matter if in use or not.
-    ///   Defaults to 65536 bytes.
-    init(readableStream: ReadableStream, bufferSize: Int = 65_536, logger: LogAgent) {
+    ///   - bridgeBufferSize: The number of bytes in the in-memory buffer.  The buffer is allocated for this size no matter if in use or not.
+    ///   Defaults to 65536 bytes (64 kb).
+    ///   - boundStreamBufferSize: The number of bytes in the buffer between the bound Foundation streams.  If `nil`, uses the
+    ///   same size as `bridgeBufferSize`.  Defaults to `nil`.  Primary use of this parameter is for testing.
+    ///   - logger: A logger that can be used to log stream events.
+    init(
+        readableStream: ReadableStream,
+        bridgeBufferSize: Int = 65_536,
+        boundStreamBufferSize: Int? = nil,
+        logger: LogAgent
+    ) {
```

Comment:
Added individual parameters for the size of the two buffers:
- The "bridge buffer", which holds data between the `ReadableStream` and the Foundation `OutputStream`, and
- the bound stream buffer, which holds data between the bound-pair `OutputStream` and `InputStream`.